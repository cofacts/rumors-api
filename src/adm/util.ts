import pino from 'pino';
import { HTTPError } from 'fets';
import type { RouterPlugin } from 'fets';
import { jwtVerify, createRemoteJWKSet, JWTPayload } from 'jose';

/** HTTP request headers to include in audit log */
const HEADERS_TO_INCLUDE = new Set([
  'cf-access-authenticated-user-email',
  'cf-connecting-ip',
  'cf-ipcountry',
]);

/**
 * JWT token content to include in audit log.
 *
 * Ref: https://developers.cloudflare.com/cloudflare-one/identity/authorization-cookie/application-token/
 * */
const TOKEN_KEYS_TO_INCLUDE = new Set([
  'aud',
  'exp',
  'iat',
  'country',
  'email', // For identity-based auth
  'common_name', // For service token auth
]);

/**
 * Verifying Cloudflare Access token.
 *
 * Ref: https://developers.cloudflare.com/cloudflare-one/identity/authorization-cookie/validating-json/#javascript-example
 */
const TEAM_DOMAIN = process.env.CLOUDFLARE_ACCESS_TEAM_DOMAIN;
const CERTS_URL = `${TEAM_DOMAIN}/cdn-cgi/access/certs`;
const JWKS = createRemoteJWKSet(new URL(CERTS_URL));

const logger = pino();

/**
 * Log request, response and user information.
 *
 * Abort request if it's not coming from Cloudflare Access.
 *
 * Ref: https://the-guild.dev/openapi/fets/server/plugins
 */
export function useAuditLog(): RouterPlugin<any, any> {
  return {
    async onRequest({ request }) {
      const requestInfo = {
        id: request.headers.get('cf-ray'),
        url: request.url,
        headers: Object.fromEntries(
          [...request.headers.entries()].filter(([key]) =>
            HEADERS_TO_INCLUDE.has(key)
          )
        ),
      };

      const token = request.headers.get('cf-access-jwt-assertion');
      if (!token) {
        throw new HTTPError(
          403,
          'Unauthorized',
          {},
          { message: 'Missing Cloudflare access token' }
        );
      }

      let payload: JWTPayload;
      try {
        payload = (
          await jwtVerify(token, JWKS, {
            issuer: TEAM_DOMAIN,
          })
        ).payload;
      } catch (e) {
        logger.error(e, 'Cloudflare access token verification failed');
        throw new HTTPError(403, 'Unauthorized', {});
      }

      const shouldIncludeBody =
        'common_name' in payload /* Called via service tokens */ ||
        request.headers.get('content-type') ===
          'application/json'; /* Probably from Swagger UI */

      logger.info(
        {
          ...requestInfo,
          user: Object.fromEntries(
            Object.entries(payload).filter(([key]) =>
              TOKEN_KEYS_TO_INCLUDE.has(key)
            )
          ),
          req: shouldIncludeBody ? await request.json() : undefined,
        },
        'Req start'
      );
    },

    async onResponse({ request, response }) {
      const shouldIncludeBody =
        response.ok &&
        (response.headers.get('content-type') ?? '').startsWith(
          'application/json'
        );

      logger.info(
        {
          id: request.headers.get('cf-ray'),
          url: request.url,
          status: response.status,
          res: shouldIncludeBody ? await response.json() : undefined,
        },
        'Req end'
      );
    },
  };
}
