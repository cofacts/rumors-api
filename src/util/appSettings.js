import yaml from 'js-yaml';
import fs from 'fs';

export class AppSettingsError extends Error {
  constructor(msg) {
    super(`[AppSettings] ${msg}`);
  }
}

export function validateYaml(yamlData) {
  if (!(yamlData instanceof Array)) {
    throw new AppSettingsError('App settings should be an array');
  }

  const dbAppIds = new Set();
  const headerAppSecrets = new Set();
  yamlData.forEach((appSetting, idx) => {
    // Check dbAppId
    //
    if (!appSetting.dbAppId)
      throw new AppSettingsError(`App setting ${idx} does not have dbAppId`);
    if (dbAppIds.has(appSetting.dbAppId))
      throw new AppSettingsError(`App setting ${idx} has duplicated dbAppId`);
    dbAppIds.add(appSetting.dbAppId);

    // Check headerAppSecrets
    //
    if (!(appSetting.headerAppSecrets instanceof Array))
      throw new AppSettingsError(
        `App setting ${idx} does not have headerAppSecrets array`
      );
    appSetting.headerAppSecrets.forEach(secret => {
      if (typeof secret !== 'string')
        throw new AppSettingsError(
          `App ${appSetting.dbAppId} has invalid secret`
        );
      if (headerAppSecrets.has(secret))
        throw new AppSettingsError(
          `App ${appSetting.dbAppId} has duplicated secret`
        );
      headerAppSecrets.add(secret);
    });
  });
}

export function loadAppSettings(path = process.env.APP_SETTINGS_PATH) {
  if (!path) throw new AppSettingsError('APP_SETTINGS_PATH is not specified');
  try {
    const yamlData = yaml.load(fs.readFileSync(path));
    validateYaml(yamlData);
    return yamlData;
  } catch (e) {
    console.error(e);
    return [];
  }
}
