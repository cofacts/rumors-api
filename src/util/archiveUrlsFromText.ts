/** Extract URLs from text and send to Internet Archive Wayback Machine */

import urlRegex from 'url-regex';
import { removeFBCLIDIfExist } from './scrapUrls';

export default async function archiveUrlsFromText(text: string) {
  const originalUrls = text.match(urlRegex()) || [];
  if (originalUrls.length === 0) return [];

  // Normalize URLs before sending to cache or scrapper to increase cache hit
  //
  const normalizedUrls = removeFBCLIDIfExist(originalUrls);

  const results = await Promise.all(
    normalizedUrls.map(async (url) => {
      const formData = new FormData();
      formData.append('url', url);
      formData.append('capture_screenshot', '1');
      formData.append('skip_first_archive', '1');
      formData.append('delay_wb_availability', '1'); // Help reduce load on IA servers

      return (
        await fetch('https://web.archive.org/save', {
          method: 'POST',
          headers: {
            Accept: 'application/json',
            Authorization: `LOW ${process.env.INTERNET_ARCHIVE_S3_ACCESS_KEY}:${process.env.INTERNET_ARCHIVE_S3_SECRET_KEY}`,
          },
          body: formData,
        })
      ).json();
    })
  );

  console.info(`[archiveUrlsFromText] Archiving ${results.length} URLs`);
  results.forEach((result, i) => {
    if (result.job_id) {
      console.info(`[archiveUrlsFromText] [ ${result.url} ]: ${result.job_id}`);
    } else {
      console.error(
        `[archiveUrlsFromText] [ ${normalizedUrls[i]} ]: ${JSON.stringify(
          result
        )}`
      );
    }
  });
  return results;
}
