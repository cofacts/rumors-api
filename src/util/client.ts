import elasticsearch from '@elastic/elasticsearch';

export default new elasticsearch.Client({
  node: process.env.ELASTICSEARCH_URL,
});

type ProcessMetaArgs<T> = {
  _id: string;
  _source: T;
  found: boolean;
  _score: number;
  highlight: object; // FIXME: use ES type
  inner_hits: object; // FIXME: use ES type
  sort: string;
  fields: object;
};

// Processes {_id, _version, found, _source: {...}} to
// {id, ..._source}.
//
export function processMeta<T extends object>({
  _id: id,
  _source: source,

  found, // for mget queries
  _score, // for search queries

  highlight, // search result highlighting. Should be {<fieldName>: ['...']}

  inner_hits, // nested query search result highlighting.

  sort, // cursor when sorted

  fields, // scripted fields (if any)
}: ProcessMetaArgs<T>) {
  if (found || _score !== undefined) {
    return {
      id,
      ...source,
      _cursor: sort,
      _score,
      highlight,
      inner_hits,
      _fields: fields,
    };
  }
  return null; // not found
}
