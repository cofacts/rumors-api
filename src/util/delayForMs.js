function delayForMs(delayMs) {
  return new Promise((resolve) => {
    setTimeout(() => resolve(), delayMs);
  });
}

export default delayForMs;
