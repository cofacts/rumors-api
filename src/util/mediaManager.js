import { MediaManager } from '@cofacts/media-manager';

export const IMAGE_PREVIEW = 'webp600w';
export const IMAGE_THUMBNAIL = 'jpg240h';

export const VIDEO_PREVIEW = 'av1-0.75x30s';
export const VIDEO_THUMBNAIL = 'av1-240p5s';

const mediaManager = new MediaManager({
  bucketName: process.env.GCS_BUCKET_NAME,
  credentialsJSON: process.env.GCS_CREDENTIALS,
  prefix: process.env.GCS_MEDIA_FOLDER,
});

export default mediaManager;
