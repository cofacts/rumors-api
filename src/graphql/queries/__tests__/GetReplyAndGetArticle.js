import gql from 'util/GraphQL';
import { loadFixtures, unloadFixtures } from 'util/fixtures';
import fixtures from '../__fixtures__/GetReplyAndArticle';
import mediaManager from 'util/mediaManager';

jest.mock('util/mediaManager');

describe('GetReplyAndGetArticle', () => {
  beforeAll(() => loadFixtures(fixtures));
  beforeEach(() => {
    mediaManager.insert.mockClear();
  });

  describe('GetArticle', () => {
    it('should get the specified article & associated replies from ID', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              text
              references {
                type
              }
              replyCount
              articleReplies(status: NORMAL) {
                status
                canUpdateStatus
                createdAt
                updatedAt
                articleId
                article {
                  id
                }
                replyId
                reply {
                  id
                  text
                  type
                  reference
                }
              }
              replyRequestCount
              replyRequests {
                reason
                feedbackCount
                positiveFeedbackCount
                negativeFeedbackCount
                article {
                  id
                }
              }
              requestedForReply
              attachmentHash
              attachmentUrl

              aiReplies {
                text
                status
              }
            }
          }
        `({}, { user: { id: 'fakeUser', appId: 'LINE' } })
      ).toMatchSnapshot();
    });

    it('fetches more than 10 reply requests', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "manyRequests") {
              text
              replyRequestCount
              replyRequests {
                reason
              }
            }
          }
        `({}, { userId: 'fakeUser', appId: 'LINE' })
      ).toMatchSnapshot();
    });

    it('should allow filtering article replies', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo3") {
              articleReplies {
                replyId
                reply {
                  id
                }
                status
              }
              normalReplies: articleReplies(status: NORMAL) {
                replyId
                reply {
                  id
                }
                status
              }
              deletedReplies: articleReplies(status: DELETED) {
                replyId
                reply {
                  id
                }
                status
              }
            }
          }
        `({}, { userId: 'fakeUser', appId: 'LINE' })
      ).toMatchSnapshot();

      // Test articleReplies's common filters
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              user1Replies: articleReplies(
                statuses: [NORMAL, DELETED]
                userId: "user1"
              ) {
                replyId
                userId
                appId
              }
              app1Replies: articleReplies(
                statuses: [NORMAL, DELETED]
                appId: "app1"
              ) {
                replyId
                userId
                appId
              }
              selfOnlyReplies: articleReplies(
                statuses: [NORMAL, DELETED]
                selfOnly: true
              ) {
                replyId
                userId
                appId
              }
            }
          }
        `({}, { userId: 'user2', appId: 'app2' })
      ).toMatchInlineSnapshot(`
        Object {
          "data": Object {
            "GetArticle": Object {
              "app1Replies": Array [
                Object {
                  "appId": "app1",
                  "replyId": "bar2",
                  "userId": "user2",
                },
                Object {
                  "appId": "app1",
                  "replyId": "bar",
                  "userId": "user1",
                },
              ],
              "selfOnlyReplies": Array [
                Object {
                  "appId": "app2",
                  "replyId": "bar3",
                  "userId": "user2",
                },
              ],
              "user1Replies": Array [
                Object {
                  "appId": "app2",
                  "replyId": "bar4",
                  "userId": "user1",
                },
                Object {
                  "appId": "app1",
                  "replyId": "bar",
                  "userId": "user1",
                },
              ],
            },
          },
        }
      `);
    });

    it('should return empty articleReply when there is none', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo2") {
              articleReplies(status: DELETED) {
                replyId
                status
              }
            }
          }
        `({}, { userId: 'fakeUser', appId: 'LINE' })
      ).toMatchSnapshot();
    });

    it('relatedArticles should work for text articles', async () => {
      // No param
      //
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              relatedArticles {
                edges {
                  cursor
                  node {
                    id
                  }
                  highlight {
                    text
                    reference
                    hyperlinks {
                      title
                      summary
                    }
                  }
                  score
                }
              }
            }
          }
        `()
      ).toMatchSnapshot('relatedArticle no-param test');

      // filter
      //
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              relatedArticles(filter: { replyCount: { GT: 0 } }) {
                edges {
                  cursor
                  node {
                    id
                    text
                  }
                  score
                }
              }
            }
          }
        `()
      ).toMatchSnapshot('relatedArticle filter test');

      // sort
      //
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              relatedArticles(orderBy: [{ _score: ASC }]) {
                edges {
                  cursor
                  node {
                    id
                    text
                  }
                  score
                }
              }
            }
          }
        `()
      ).toMatchSnapshot('relatedArticle sorting test');
    });

    it('realtedArticles should work for media articles', async () => {
      mediaManager.query.mockImplementationOnce(async () => ({
        queryInfo: {
          type: 'image',
          id: fixtures['/articles/doc/mediaArticle'].attachmentHash,
        },
        hits: [
          {
            similarity: 1,
            entry: {
              id: fixtures['/articles/doc/mediaArticle'].attachmentHash,
            },
          },
          {
            // Simulate the case when Google drive has a media entry,
            // but no article is in Elasticsearch has that attachmentHash
            similarity: 0.9,
            entry: { id: 'exist-on-google-drive-but-not-in-db' },
          },
          {
            similarity: 0.75,
            entry: {
              id: fixtures['/articles/doc/similarMediaArticle'].attachmentHash,
            },
          },
        ],
      }));

      expect(
        await gql`
          {
            GetArticle(id: "mediaArticle") {
              relatedArticles {
                edges {
                  cursor
                  node {
                    id
                  }
                  score
                }
              }
            }
          }
        `()
      ).toMatchSnapshot();

      // Expect called exactly once with {id: ...}
      //
      expect(mediaManager.query.mock.calls).toMatchInlineSnapshot(`
        Array [
          Array [
            Object {
              "id": "hash-for-media-article",
            },
          ],
        ]
      `);
    });

    it('relatedArticle of blocked article can include both normal and blocked articles', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "blockedArticle1") {
              relatedArticles {
                edges {
                  node {
                    id
                  }
                }
              }
            }
          }
        `()
      ).toMatchInlineSnapshot(`
        Object {
          "data": Object {
            "GetArticle": Object {
              "relatedArticles": Object {
                "edges": Array [
                  Object {
                    "node": Object {
                      "id": "foo3",
                    },
                  },
                  Object {
                    "node": Object {
                      "id": "foo2",
                    },
                  },
                  Object {
                    "node": Object {
                      "id": "blockedArticle2",
                    },
                  },
                  Object {
                    "node": Object {
                      "id": "mediaArticle",
                    },
                  },
                  Object {
                    "node": Object {
                      "id": "foo",
                    },
                  },
                ],
              },
            },
          },
        }
      `);
    });

    it('feedbacks should work', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              articleReplies {
                feedbacks {
                  id
                  vote
                }
                positiveFeedbackCount
                negativeFeedbackCount
                feedbackCount
                createdAt
              }
            }
          }
        `()
      ).toMatchSnapshot('feedback loading test');
    });

    it('get specified article and articleCategories with NORMAL status', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              text
              categoryCount
              articleCategories {
                id
                categoryId
                category {
                  id
                  title
                  description
                }
              }
            }
          }
        `()
      ).toMatchSnapshot();
    });

    it('get specified article and articleCategories with DELETED status', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              text
              articleCategories(status: DELETED) {
                categoryId
                category {
                  id
                  title
                  description
                }
              }
            }
          }
        `()
      ).toMatchSnapshot();
    });

    it('authenticated fields returns null for non-logged users', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "foo") {
              requestedForReply
            }
          }
        `({}, { appId: 'LINE' })
      ).toMatchSnapshot();
    });

    it('can list AI Transcripts', async () => {
      expect(
        await gql`
          {
            GetArticle(id: "mediaArticle") {
              id
              aiTranscripts {
                text
                status
              }
            }
          }
        `()
      ).toMatchInlineSnapshot(`
        Object {
          "data": Object {
            "GetArticle": Object {
              "aiTranscripts": Array [
                Object {
                  "status": "SUCCESS",
                  "text": "Lorum ipsum",
                },
              ],
              "id": "mediaArticle",
            },
          },
        }
      `);
    });
  });

  describe('GetReply', () => {
    it('should get the specified reply & associated articles from ID', async () => {
      expect(
        await gql`
          {
            GetReply(id: "bar") {
              text
              type
              reference
              articleReplies {
                canUpdateStatus
                articleId
                article {
                  text
                }
              }
              status
            }
          }
        `()
      ).toMatchSnapshot();
    });

    it('should allow filtering article replies', async () => {
      expect(
        await gql`
          {
            GetReply(id: "bar2") {
              articleReplies {
                articleId
                article {
                  id
                }
                status
              }
              normalReplies: articleReplies(status: NORMAL) {
                articleId
                article {
                  id
                }
                status
              }
              deletedReplies: articleReplies(status: DELETED) {
                articleId
                article {
                  id
                }
                status
              }
            }
          }
        `()
      ).toMatchSnapshot();
    });
    it('similarReplies should work', async () => {
      // No param
      //
      expect(
        await gql`
          {
            GetReply(id: "bar") {
              similarReplies {
                edges {
                  cursor
                  node {
                    id
                  }
                  score
                  highlight {
                    text
                    reference
                    hyperlinks {
                      title
                      summary
                    }
                  }
                }
              }
            }
          }
        `()
      ).toMatchSnapshot('similarReply no-param test');

      // sort
      //
      expect(
        await gql`
          {
            GetReply(id: "bar") {
              similarReplies(orderBy: [{ _score: ASC }]) {
                edges {
                  cursor
                  node {
                    id
                  }
                  score
                  highlight {
                    text
                    reference
                    hyperlinks {
                      title
                      summary
                    }
                  }
                }
              }
            }
          }
        `()
      ).toMatchSnapshot('similarReply sorting test');
    });

    it('sets status to BLOCKED when the author is blocked', async () => {
      expect(
        await gql`
          {
            GetReply(id: "bar5") {
              status
            }
          }
        `()
      ).toMatchInlineSnapshot(`
        Object {
          "data": Object {
            "GetReply": Object {
              "status": "BLOCKED",
            },
          },
        }
      `);
    });
  });

  afterAll(() => unloadFixtures(fixtures));
});
