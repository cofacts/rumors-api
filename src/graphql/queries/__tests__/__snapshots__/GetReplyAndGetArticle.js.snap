// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetReplyAndGetArticle GetArticle authenticated fields returns null for non-logged users 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "requestedForReply": null,
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle feedbacks should work: feedback loading test 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "articleReplies": Array [
        Object {
          "createdAt": "2015-01-01T12:10:30Z",
          "feedbackCount": 1,
          "feedbacks": Array [
            Object {
              "id": "foo__bar__test-user__test-app",
              "vote": "UPVOTE",
            },
          ],
          "negativeFeedbackCount": 0,
          "positiveFeedbackCount": 1,
        },
      ],
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle fetches more than 10 reply requests 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "replyRequestCount": 11,
      "replyRequests": Array [
        Object {
          "reason": "Reason 0",
        },
        Object {
          "reason": "Reason 1",
        },
        Object {
          "reason": "Reason 2",
        },
        Object {
          "reason": "Reason 3",
        },
        Object {
          "reason": "Reason 4",
        },
        Object {
          "reason": "Reason 5",
        },
        Object {
          "reason": "Reason 6",
        },
        Object {
          "reason": "Reason 7",
        },
        Object {
          "reason": "Reason 8",
        },
        Object {
          "reason": "Reason 9",
        },
        Object {
          "reason": "Reason 10",
        },
      ],
      "text": "Popular",
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle get specified article and articleCategories with DELETED status 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "articleCategories": Array [
        Object {
          "category": Object {
            "description": "詐騙貼圖、假行銷手法。",
            "id": "c2",
            "title": "免費訊息詐騙",
          },
          "categoryId": "c2",
        },
      ],
      "text": "Lorum ipsum",
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle get specified article and articleCategories with NORMAL status 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "articleCategories": Array [
        Object {
          "category": Object {
            "description": "對同性婚姻的恐懼、愛滋病的誤解與的防疫相關釋疑。",
            "id": "c1",
            "title": "性少數與愛滋病",
          },
          "categoryId": "c1",
          "id": "foo__c1",
        },
      ],
      "categoryCount": 1,
      "text": "Lorum ipsum",
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle realtedArticles should work for media articles 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "relatedArticles": Object {
        "edges": Array [
          Object {
            "cursor": "WzcuNSwic2ltaWxhck1lZGlhQXJ0aWNsZSJd",
            "node": Object {
              "id": "similarMediaArticle",
            },
            "score": 7.5,
          },
          Object {
            "cursor": "WzAuNTQ2MTgxODYsImZvbzMiXQ==",
            "node": Object {
              "id": "foo3",
            },
            "score": 0.54618186,
          },
          Object {
            "cursor": "WzAuNTMwMzI5NiwiZm9vMiJd",
            "node": Object {
              "id": "foo2",
            },
            "score": 0.5303296,
          },
          Object {
            "cursor": "WzAuNDg3ODUxNywiZm9vIl0=",
            "node": Object {
              "id": "foo",
            },
            "score": 0.4878517,
          },
        ],
      },
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle relatedArticles should work for text articles: relatedArticle filter test 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "relatedArticles": Object {
        "edges": Array [
          Object {
            "cursor": "WzAuOTg2NDk1MywiZm9vMyJd",
            "node": Object {
              "id": "foo3",
              "text": "Lorum ipsum Lorum ipsum Lorum ipsum",
            },
            "score": 0.9864953,
          },
          Object {
            "cursor": "WzAuNTMwMzI5NiwiZm9vMiJd",
            "node": Object {
              "id": "foo2",
              "text": "Lorum ipsum Lorum ipsum",
            },
            "score": 0.5303296,
          },
        ],
      },
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle relatedArticles should work for text articles: relatedArticle no-param test 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "relatedArticles": Object {
        "edges": Array [
          Object {
            "cursor": "WzAuOTg2NDk1MywiZm9vMyJd",
            "highlight": Object {
              "hyperlinks": Array [
                Object {
                  "summary": "<HIGHLIGHT>summary</HIGHLIGHT>",
                  "title": "<HIGHLIGHT>title</HIGHLIGHT> <HIGHLIGHT>title</HIGHLIGHT>",
                },
              ],
              "reference": null,
              "text": "<HIGHLIGHT>Lorum</HIGHLIGHT> <HIGHLIGHT>ipsum</HIGHLIGHT> <HIGHLIGHT>Lorum</HIGHLIGHT> <HIGHLIGHT>ipsum</HIGHLIGHT> <HIGHLIGHT>Lorum</HIGHLIGHT> <HIGHLIGHT>ipsum</HIGHLIGHT>",
            },
            "node": Object {
              "id": "foo3",
            },
            "score": 0.9864953,
          },
          Object {
            "cursor": "WzAuNTMwMzI5NiwiZm9vMiJd",
            "highlight": Object {
              "hyperlinks": Array [],
              "reference": null,
              "text": "<HIGHLIGHT>Lorum</HIGHLIGHT> <HIGHLIGHT>ipsum</HIGHLIGHT> <HIGHLIGHT>Lorum</HIGHLIGHT> <HIGHLIGHT>ipsum</HIGHLIGHT>",
            },
            "node": Object {
              "id": "foo2",
            },
            "score": 0.5303296,
          },
          Object {
            "cursor": "WzAuNDg3ODUxNywibWVkaWFBcnRpY2xlIl0=",
            "highlight": Object {
              "hyperlinks": Array [],
              "reference": null,
              "text": "<HIGHLIGHT>Lorum</HIGHLIGHT> <HIGHLIGHT>ipsum</HIGHLIGHT>",
            },
            "node": Object {
              "id": "mediaArticle",
            },
            "score": 0.4878517,
          },
        ],
      },
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle relatedArticles should work for text articles: relatedArticle sorting test 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "relatedArticles": Object {
        "edges": Array [
          Object {
            "cursor": "WzAuNDg3ODUxNywibWVkaWFBcnRpY2xlIl0=",
            "node": Object {
              "id": "mediaArticle",
              "text": "Lorum ipsum",
            },
            "score": 0.4878517,
          },
          Object {
            "cursor": "WzAuNTMwMzI5NiwiZm9vMiJd",
            "node": Object {
              "id": "foo2",
              "text": "Lorum ipsum Lorum ipsum",
            },
            "score": 0.5303296,
          },
          Object {
            "cursor": "WzAuOTg2NDk1MywiZm9vMyJd",
            "node": Object {
              "id": "foo3",
              "text": "Lorum ipsum Lorum ipsum Lorum ipsum",
            },
            "score": 0.9864953,
          },
        ],
      },
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle should allow filtering article replies 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "articleReplies": Array [
        Object {
          "reply": Object {
            "id": "fofo",
          },
          "replyId": "fofo",
          "status": "NORMAL",
        },
      ],
      "deletedReplies": Array [
        Object {
          "reply": Object {
            "id": "bar2",
          },
          "replyId": "bar2",
          "status": "DELETED",
        },
      ],
      "normalReplies": Array [
        Object {
          "reply": Object {
            "id": "fofo",
          },
          "replyId": "fofo",
          "status": "NORMAL",
        },
      ],
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle should get the specified article & associated replies from ID 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "aiReplies": Array [
        Object {
          "status": "SUCCESS",
          "text": "AI says we should be careful of this message",
        },
      ],
      "articleReplies": Array [
        Object {
          "article": Object {
            "id": "foo",
          },
          "articleId": "foo",
          "canUpdateStatus": false,
          "createdAt": "2015-01-01T12:10:30Z",
          "reply": Object {
            "id": "bar",
            "reference": "barbar",
            "text": "bar",
            "type": "NOT_ARTICLE",
          },
          "replyId": "bar",
          "status": "NORMAL",
          "updatedAt": "2015-01-02T12:10:30Z",
        },
      ],
      "attachmentHash": null,
      "attachmentUrl": null,
      "references": Array [
        Object {
          "type": "LINE",
        },
      ],
      "replyCount": 1,
      "replyRequestCount": 2,
      "replyRequests": Array [
        Object {
          "article": Object {
            "id": "foo",
          },
          "feedbackCount": 2,
          "negativeFeedbackCount": 1,
          "positiveFeedbackCount": 1,
          "reason": "Reason foo",
        },
        Object {
          "article": Object {
            "id": "foo",
          },
          "feedbackCount": 0,
          "negativeFeedbackCount": 0,
          "positiveFeedbackCount": 0,
          "reason": null,
        },
      ],
      "requestedForReply": true,
      "text": "Lorum ipsum",
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetArticle should return empty articleReply when there is none 1`] = `
Object {
  "data": Object {
    "GetArticle": Object {
      "articleReplies": Array [],
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetReply should allow filtering article replies 1`] = `
Object {
  "data": Object {
    "GetReply": Object {
      "articleReplies": Array [
        Object {
          "article": Object {
            "id": "foo2",
          },
          "articleId": "foo2",
          "status": "NORMAL",
        },
      ],
      "deletedReplies": Array [
        Object {
          "article": Object {
            "id": "foo",
          },
          "articleId": "foo",
          "status": "DELETED",
        },
        Object {
          "article": Object {
            "id": "foo3",
          },
          "articleId": "foo3",
          "status": "DELETED",
        },
      ],
      "normalReplies": Array [
        Object {
          "article": Object {
            "id": "foo2",
          },
          "articleId": "foo2",
          "status": "NORMAL",
        },
      ],
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetReply should get the specified reply & associated articles from ID 1`] = `
Object {
  "data": Object {
    "GetReply": Object {
      "articleReplies": Array [
        Object {
          "article": Object {
            "text": "Lorum ipsum",
          },
          "articleId": "foo",
          "canUpdateStatus": false,
        },
      ],
      "reference": "barbar",
      "status": "NORMAL",
      "text": "bar",
      "type": "NOT_ARTICLE",
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetReply similarReplies should work: similarReply no-param test 1`] = `
Object {
  "data": Object {
    "GetReply": Object {
      "similarReplies": Object {
        "edges": Array [
          Object {
            "cursor": "WzIuMTE2NjU0LCJzaW1pbGFyLXRvLWJhciJd",
            "highlight": Object {
              "hyperlinks": Array [],
              "reference": "<HIGHLIGHT>barbar</HIGHLIGHT>",
              "text": "<HIGHLIGHT>bar</HIGHLIGHT> <HIGHLIGHT>bar</HIGHLIGHT>",
            },
            "node": Object {
              "id": "similar-to-bar",
            },
            "score": 2.116654,
          },
          Object {
            "cursor": "WzAuNTY2MzE3Miwic2ltaWxhci10by1iYXIyIl0=",
            "highlight": Object {
              "hyperlinks": Array [
                Object {
                  "summary": "<HIGHLIGHT>Lorem</HIGHLIGHT> <HIGHLIGHT>Ipsum</HIGHLIGHT> <HIGHLIGHT>Ipsum</HIGHLIGHT>",
                  "title": "<HIGHLIGHT>GG</HIGHLIGHT> G",
                },
              ],
              "reference": null,
              "text": null,
            },
            "node": Object {
              "id": "similar-to-bar2",
            },
            "score": 0.5663172,
          },
        ],
      },
    },
  },
}
`;

exports[`GetReplyAndGetArticle GetReply similarReplies should work: similarReply sorting test 1`] = `
Object {
  "data": Object {
    "GetReply": Object {
      "similarReplies": Object {
        "edges": Array [
          Object {
            "cursor": "WzAuNTY2MzE3Miwic2ltaWxhci10by1iYXIyIl0=",
            "highlight": Object {
              "hyperlinks": Array [
                Object {
                  "summary": "<HIGHLIGHT>Lorem</HIGHLIGHT> <HIGHLIGHT>Ipsum</HIGHLIGHT> <HIGHLIGHT>Ipsum</HIGHLIGHT>",
                  "title": "<HIGHLIGHT>GG</HIGHLIGHT> G",
                },
              ],
              "reference": null,
              "text": null,
            },
            "node": Object {
              "id": "similar-to-bar2",
            },
            "score": 0.5663172,
          },
          Object {
            "cursor": "WzIuMTE2NjU0LCJzaW1pbGFyLXRvLWJhciJd",
            "highlight": Object {
              "hyperlinks": Array [],
              "reference": "<HIGHLIGHT>barbar</HIGHLIGHT>",
              "text": "<HIGHLIGHT>bar</HIGHLIGHT> <HIGHLIGHT>bar</HIGHLIGHT>",
            },
            "node": Object {
              "id": "similar-to-bar",
            },
            "score": 2.116654,
          },
        ],
      },
    },
  },
}
`;
